Language server

BarryNolte opened the pull request:
# New items in the PR
Resolves #51 
## Project Layout

To facilitate a better edit/build/debug process, the extension was moved to a 'client' directory, the language server to a 'server' directory. **_Note:_** This may break the nightly build process and will need to be adjusted.

## Language Server

1. Identifies links and imports so the editor can mark
   them as such. Ctrl+click on the links will attempt
   to launch what the link points to.
2. 'Find All References' will show a list (as presented
   by vscode) of all references of the given node.
3. 'Goto Definition' will do one of two things. If there
   are multiple definitions, it will show a list similar
   to 'Find All References'. If there is only one option
   for the definition, then the cursor will jump to that
   option.
4. 'Rename...' will rename a node. If there are multiple
   nodes (a.style... / a.shape... / a.label... / etc.),
   then a rename will change all locations that contain
   that node.
5. Auto Completion:

    a) Typing a '@' will show a list of d2 files that are contained in the d2 documents current folder and all child folders from there.
    
    b) Typing a '.' that is along a valid path, will bring up the proper list for that position in the path.
    
    c) Typing a ':' will bring up a list of possible values for that position in the current path. If there is no list, in the case of a free-form field such as 'label', then no list is shown.

    d) The 'Tab' key will commit the auto completion.
6. When a preview is not present, then there is no error messages generated, the language server will show any errors that are created during the AST phase of d2 compilation.

alixander commented:
(apologies for the delay on this, team is super busy with some d2 studio stuff, very looking forward to digging into this and trying it soon)

gavin-ts commented on .vscode/settings.json:
```diff
@@ -0,0 +1,5 @@
+{
```
isn't this a personal config file? maybe add .vscode to .gitignore instead?

BarryNolte replied:
Yes, removed.

gavin-ts commented on client/src/docToPreviewGenerator.ts:
```diff
@@ -89,7 +89,13 @@ export class DocToPreviewGenerator {
         arr.forEach((s) => {
-          list += `<li>${s}</li>`;
+          if (s.length > 0) {
+            const rg = new RegExp(/^(.*?):(\d+):(\d+):(\s+)(.*)$/g).exec(s);
```
I think you can create the regexp once outside and use the same one in the forEach
Adding a comment with an example match would be helpful

BarryNolte replied:
For readability I'd prefer to keep this all together, this loop will be execute 3 or less times in the 99.9% case.  Performance is not an issue here. I did add a couple of examples.

gavin-ts commented on package.json:
```diff
@@ -320,9 +323,6 @@
     "webpack-cli": "^4.10.0"
   },
-  "dependencies": {
-    "markdown-it-container": "^3.0.0"
```
do we no longer need this?

BarryNolte replied:
This has been moved to the clients package.json

gavin-ts commented on server/package.json:
```diff
@@ -0,0 +1,38 @@
+  },
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/Microsoft/vscode-extension-samples"
```
https://github.com/terrastruct/d2-vscode

BarryNolte replied:
Good catch.

gavin-ts reviewed (commented):
Note: we should add `yarn installdep` as a first step to the readme

ran into this issue when using an older d2 version, looks like it has errors trying to parse the svg response: 

<img width="455" alt="Screenshot 2023-12-15 at 8 09 56 PM" src="https://github.com/terrastruct/d2-vscode/assets/85081687/32a419f0-57ba-485a-a5ae-96df6ba5e1d9">

```
***************************
D2 Language Server Starting

Client:  Visual Studio Code
Version: 1.85.0
PID:     11423
***************************

[Info  - 8:09:44 PM] Language Server D2 path: d2
SyntaxError: Unexpected token '<', "<?xml vers"... is not valid JSON
    at JSON.parse (<anonymous>)
    at new AstReader (.../terrastruct.d2-0.8.7/server/dist/d2Ast.js:18:28)
    at .../terrastruct.d2-0.8.7/server/dist/server.js:97:15
    at CallbackList.invoke (.../terrastruct.d2-0.8.7/server/node_modules/vscode-jsonrpc/lib/common/events.js:55:39)
    at Emitter.fire (.../terrastruct.d2-0.8.7/server/node_modules/vscode-jsonrpc/lib/common/events.js:117:36)
    at .../terrastruct.d2-0.8.7/server/node_modules/vscode-languageserver/lib/common/textDocuments.js:122:38
    at handleNotification (.../terrastruct.d2-0.8.7/server/node_modules/vscode-jsonrpc/lib/common/connection.js:640:25)
    at handleMessage (.../terrastruct.d2-0.8.7/server/node_modules/vscode-jsonrpc/lib/common/connection.js:342:13)
    at processMessageQueue (.../terrastruct.d2-0.8.7/server/node_modules/vscode-jsonrpc/lib/common/connection.js:362:17)
    at Immediate._onImmediate (.../terrastruct.d2-0.8.7/server/node_modules/vscode-jsonrpc/lib/common/connection.js:334:13)
[Error - 8:09:47 PM] Request textDocument/documentLink failed.
  Message: Request textDocument/documentLink failed with message: Cannot read properties of undefined (reading 'LinksAndImports')
  Code: -32603 
```

gavin-ts commented:
go to definition worked for objects at the outermost level but doesn't seem to work for nested objects

<img width="230" alt="Screenshot 2023-12-15 at 8 33 12 PM" src="https://github.com/terrastruct/d2-vscode/assets/85081687/8c375556-b905-4f4a-8aa1-6558e89b5f87">

gavin-ts commented on server/src/server.ts:
```diff
@@ -0,0 +1,314 @@
+    let docPath = link.str.replace(/['|"]/g, "");
+
+    // If '.d2' is there, get rid of it.
+    docPath = path.basename(docPath, ".d2");
```
this function should check what the extension is, if it is not d2, this breaks the link to it (and tries to go to file.txt.d2 instead.

gavin-ts commented:
:100:

<img width="483" alt="Screenshot 2023-12-15 at 8 40 42 PM" src="https://github.com/terrastruct/d2-vscode/assets/85081687/cf850491-12e0-4af8-bbb1-04ae82fbe10f">

gavin-ts commented:
I wonder if this can check for any d2 files inside the workspace folder rather than just the current `.d2` file's folder? currently it doesn't seem to be able to lookup anything `../etc` even if it is open within the workspace

gavin-ts commented:
`@` and `.` work great but I couldn't get `:` to work as described

`.` 💯 
<img width="534" alt="Screenshot 2023-12-15 at 8 44 52 PM" src="https://github.com/terrastruct/d2-vscode/assets/85081687/c7ef679b-c33e-497b-909d-fa7f5afd2a22">

`:`
<img width="627" alt="Screenshot 2023-12-15 at 8 44 42 PM" src="https://github.com/terrastruct/d2-vscode/assets/85081687/f90f24fc-481b-4a6d-a424-fba4e46527e0">


It would be nice to have children appear here too but it could be a separate issue:
<img width="477" alt="Screenshot 2023-12-15 at 8 46 41 PM" src="https://github.com/terrastruct/d2-vscode/assets/85081687/48eb86eb-1bcd-4b08-ae08-63bff8d10f1c">

BarryNolte commented:
The exception that was firing on the JSON parse was due to the refactoring of the D2 executable code.  When a parse happens, it should be allowed to, "fail", so that the ast and errors can be picked up and transmitted to the extension.  The parser being fault tolerant allows for incomplete d2 code to be parsed and that gives enough information for autocomplete to work. The code was just returning because there was an error, (which happens about every fifth character you type as you create a diagram), and that exception was that the ast was empty.  I've added a little more error checking around that scenario and have a PR ready for the D2 code tree.

BarryNolte commented:
Workspaces are now honored for d2 file search.  Nested objects work correctly.  I took out my, 'I know better', code from the ':' path for autocomplete. Now we should be at the point of, this works, but what else would be cool to add?

MarvinJWendt commented:
I just want to say I already love this PR so much. This is literally the only feature that I miss in D2 (and my most important selling point). Thank you, @BarryNolte, for working on this.

> Now we should be at the point of, this works, but what else would be cool to add?

What do you (all) think about releasing a beta of the LSP + VSC Extension? Things that are "cool to add" are probably worth looking into, when the LSP proved to be stable enough. Most cool things will show when the community uses it and finds special use cases. This would probably also decrease the time needed, as features could be prioritized by demand.

(Just my 2 cents)
Cheers

BarryNolte commented:
> I just want to say I already love this PR so much. This is literally the only feature that I miss in D2 (and my most important selling point). Thank you, @BarryNolte, for working on this.

Thanks @MarvinJWendt!  This has been fun to work on.  It's a far cry from my previous work on performance tools and debuggers.  This should be available once the team at Terrastruct accepts the PR.

BarryNolte commented:
@alixander and @gavin-ts Ping.

alixander reviewed (commented):
@BarryNolte I'm running into this error when I try to install. Any ideas?

Error at the end: `ERROR  Extension entrypoint(s) missing.`

```sh
~/dev/d2-vscode  LanguageServer ✔                                                                                                                                                                                                                    31d22h  ⍉ 31d22h  ⍉
▶ npm run dev

> d2@0.8.8 dev
> code --uninstall-extension terrastruct.d2; yarn gen && yarn pkg && code --install-extension d2.vsix

Extension 'terrastruct.d2' is not installed.
Make sure you use the full extension ID, including the publisher, e.g.: ms-dotnettools.csharp
yarn run v1.22.21
warning d2@0.8.8: The engine "vscode" appears to be invalid.
$ yq -o=json . syntaxes/d2.tmLanguage.yaml > syntaxes/d2.tmLanguage.json
✨  Done in 0.04s.
yarn run v1.22.21
warning d2@0.8.8: The engine "vscode" appears to be invalid.
$ vsce package --out d2.vsix
(node:99072) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
 INFO  Detected presence of yarn.lock. Using 'yarn' instead of 'npm' (to override this pass '--no-yarn' on the command line).
Executing prepublish script 'yarn run vscode:prepublish'...
warning d2@0.8.8: The engine "vscode" appears to be invalid.
$ yarn run package
warning d2@0.8.8: The engine "vscode" appears to be invalid.
$ yarn run package_client && yarn run package_server
warning d2@0.8.8: The engine "vscode" appears to be invalid.
$ webpack --config ./client/webpack.config.js --mode production --devtool hidden-source-map
    [webpack-cli] Compiler starting...
    [webpack-cli] Compiler is using config: '/Users/nca/dev/d2-vscode/client/webpack.config.js'
asset extension.js 362 KiB [compared for emit] [minimized] (name: main) 1 related asset
asset ../client/dist/utility.d.ts 1.17 KiB [compared for emit]
asset ../client/dist/extension.d.ts 880 bytes [compared for emit]
asset ../client/dist/docToPreviewGenerator.d.ts 814 bytes [compared for emit]
asset ../client/dist/browserWindow.d.ts 603 bytes [compared for emit]
asset ../client/dist/refreshTimer.d.ts 527 bytes [compared for emit]
asset ../client/dist/outputChannel.d.ts 453 bytes [compared for emit]
asset ../client/dist/taskRunner.d.ts 453 bytes [compared for emit]
asset ../client/dist/tasks.d.ts 430 bytes [compared for emit]
asset ../client/dist/themePicker.d.ts 411 bytes [compared for emit]
asset ../client/dist/layoutPicker.d.ts 251 bytes [compared for emit]
runtime modules 88 bytes 1 module
modules by path ./client/node_modules/ 890 KiB 129 modules
modules by path ./client/src/*.ts 33.7 KiB
  ./client/src/extension.ts 9.97 KiB [built] [code generated]
  ./client/src/docToPreviewGenerator.ts 3.55 KiB [built] [code generated]
  ./client/src/utility.ts 3.95 KiB [built] [code generated]
  ./client/src/outputChannel.ts 1.18 KiB [built] [code generated]
  ./client/src/layoutPicker.ts 1.6 KiB [built] [code generated]
  + 5 modules
+ 8 modules

WARNING in ./client/node_modules/vscode-languageserver-types/lib/umd/main.js 3:24-31
Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
 @ ./client/node_modules/vscode-languageserver-protocol/lib/common/api.js 23:13-51
 @ ./client/node_modules/vscode-languageserver-protocol/lib/node/main.js 24:13-37
 @ ./client/node_modules/vscode-languageclient/lib/common/api.js 22:13-54
 @ ./client/node_modules/vscode-languageclient/lib/node/main.js 34:13-37
 @ ./client/node_modules/vscode-languageclient/node.js 7:0-43
 @ ./client/src/extension.ts 18:15-52

1 warning has detailed information that is not shown.
Use 'stats.errorDetails: true' resp. '--stats-error-details' to show it.

webpack 5.88.2 compiled with 1 warning in 2940 ms
    [webpack-cli] Compiler finished
warning d2@0.8.8: The engine "vscode" appears to be invalid.
$ webpack --config ./server/webpack.config.js --mode production --devtool hidden-source-map
    [webpack-cli] Compiler starting...
    [webpack-cli] Compiler is using config: '/Users/nca/dev/d2-vscode/server/webpack.config.js'
asset server.js 200 KiB [compared for emit] [minimized] (name: main) 1 related asset
asset ../server/dist/dataContainers.d.ts 4.21 KiB [compared for emit]
asset ../server/dist/d2Ast.d.ts 2.06 KiB [compared for emit]
asset ../server/dist/completionTree.d.ts 1.38 KiB [compared for emit]
asset ../server/dist/server.d.ts 763 bytes [compared for emit]
asset ../server/dist/completionHelpers.d.ts 649 bytes [compared for emit]
asset ../server/dist/utility.d.ts 311 bytes [compared for emit]
runtime modules 88 bytes 1 module
modules by path ./server/node_modules/ 469 KiB
  cacheable modules 468 KiB 72 modules
  ./server/node_modules/vscode-languageserver-textdocument/lib/umd/ sync 160 bytes [built] [code generated]
  ./server/node_modules/vscode-languageserver-types/lib/umd/ sync 160 bytes [built] [code generated]
modules by path ./server/src/*.ts 46.2 KiB
  ./server/src/server.ts 8.19 KiB [built] [code generated]
  ./server/src/d2Ast.ts 5.42 KiB [built] [code generated]
  ./server/src/completionHelpers.ts 5.06 KiB [built] [code generated]
  ./server/src/dataContainers.ts 14.1 KiB [built] [code generated]
  ./server/src/utility.ts 1.82 KiB [built] [code generated]
  ./server/src/completionTree.ts 11.6 KiB [built] [code generated]
+ 10 modules

WARNING in ./server/node_modules/vscode-languageserver-textdocument/lib/umd/main.js 12:24-31
Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
 @ ./server/src/server.ts 10:45-90

WARNING in ./server/node_modules/vscode-languageserver-types/lib/umd/main.js 3:24-31
Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
 @ ./server/node_modules/vscode-languageserver-protocol/lib/common/api.js 23:13-51
 @ ./server/node_modules/vscode-languageserver-protocol/lib/node/main.js 24:13-37
 @ ./server/node_modules/vscode-languageserver/lib/common/api.js 25:13-55
 @ ./server/node_modules/vscode-languageserver/lib/node/main.js 29:13-37
 @ ./server/src/completionHelpers.ts 7:32-64
 @ ./server/src/server.ts 13:28-58

2 warnings have detailed information that is not shown.
Use 'stats.errorDetails: true' resp. '--stats-error-details' to show it.

webpack 5.88.2 compiled with 2 warnings in 2339 ms
    [webpack-cli] Compiler finished
 INFO  Detected presence of yarn.lock. Using 'yarn' instead of 'npm' (to override this pass '--no-yarn' on the command line).
 ERROR  Extension entrypoint(s) missing. Make sure these files exist and aren't ignored by '.vscodeignore':
  extension/client/dist/extension.js
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

BarryNolte commented:
> @BarryNolte I'm running into this error when I try to install. Any ideas?
> 
> Error at the end: `ERROR Extension entrypoint(s) missing.`
> 

I moved a ton of files around to separate the client from the server, so this could be leftovers from the previous directory structure.  I'd delete the whole directory, re-clone the repository and build from there.  As I moved things around, I ran into numerous issues just like this that were fixed by deep cleaning the node modules directories and deleting ./dist/... directories.

alixander reviewed (commented):
from a completely fresh clone to the branch, i still get the error.

can you try these commands as well?

```
/dev
▶ git clone -b LanguageServer git@github.com:BarryNolte/d2-vscode.git barry-vscode
Cloning into 'barry-vscode'...
remote: Enumerating objects: 1323, done.
remote: Counting objects: 100% (775/775), done.
remote: Compressing objects: 100% (292/292), done.
remote: Total 1323 (delta 540), reused 676 (delta 462), pack-reused 548
Receiving objects: 100% (1323/1323), 5.58 MiB | 4.00 MiB/s, done.
Resolving deltas: 100% (791/791), done.

~/dev
▶ cd barry-vscode

~/dev/barry-vscode  LanguageServer ✔                                                                                                                                                                                                                                          35d0h   35d0h
▶ yarn installdep
yarn run v1.22.21
warning d2@0.8.8: The engine "vscode" appears to be invalid.
$ yarn install && cd client && yarn install && cd ../server && yarn install && cd ..
[1/5] 🔍  Validating package.json...
warning d2@0.8.8: The engine "vscode" appears to be invalid.
[2/5] 🔍  Resolving packages...
[3/5] 🚚  Fetching packages...
[4/5] 🔗  Linking dependencies...
[5/5] 🔨  Building fresh packages...
[1/5] 🔍  Validating package.json...
warning d2-extension-client@0.8.7: The engine "vscode" appears to be invalid.
[2/5] 🔍  Resolving packages...
[3/5] 🚚  Fetching packages...
warning Pattern ["wrap-ansi@^7.0.0"] is trying to unpack in the same destination "/Users/nca/Library/Caches/Yarn/v6/npm-wrap-ansi-cjs-7.0.0-integrity/node_modules/wrap-ansi-cjs" as pattern ["wrap-ansi-cjs@npm:wrap-ansi@^7.0.0"]. This could result in non-deterministic behavior, skipping
.
warning vscode-languageclient@9.0.1: The engine "vscode" appears to be invalid.
[4/5] 🔗  Linking dependencies...
[5/5] 🔨  Building fresh packages...
[1/5] 🔍  Validating package.json...
[2/5] 🔍  Resolving packages...
[3/5] 🚚  Fetching packages...
[4/5] 🔗  Linking dependencies...
[5/5] 🔨  Building fresh packages...
✨  Done in 4.52s.

~/dev/barry-vscode  LanguageServer ✔                                                                                                                                                                                                                                          35d0h   35d0h
▶ npm run dev

...rest of output is same as above
```

alixander commented:
it must be that it's building as `.ts` and vscode is looking for `.js`?

BarryNolte reviewed (commented):
It was looking for .js files, then just weren't being generated.  I updated package.json to fix that.  Since I run "npm run watch", it continuously compiles and I didn't notice it not happening when I ran "npm run dev", since everything was already compiled.  Should work now.

alixander commented on make.sh:
```diff
@@ -19,5 +19,6 @@ runjob fmt ./ci/sub/bin/fmt.sh &
   runjob lint &
 fi
-runjob build 'yarn package' &
+runjob buildinit 'yarn installdep'
+runjob build 'yarn package'
```
if you don't add the `&`, I think it won't run the jobs concurrently

I think it should be like (but i'm not sure)
`runjob (buildinit 'yarn installdep' && build 'yarn package') &`

alixander commented on server/src/d2Ast.ts:
```diff
@@ -0,0 +1,202 @@
+    try {
+      this.d2Info = JSON.parse(astStr);
+    } catch (err) {
+      connection.console.error("Possilble D2/VsCode version missmatch.");
```
typos

alixander commented on server/src/d2Ast.ts:
```diff
@@ -0,0 +1,202 @@
+
+    this.doNodes(this.d2Info?.Ast.nodes);
+
+    // console.info(JSON.stringify(this.d2Info, null, 2));
```
dev comment

alixander commented on server/src/d2Ast.ts:
```diff
@@ -0,0 +1,202 @@
+  get Errors(): PublishDiagnosticsParams | undefined {
+    const diags: Diagnostic[] = [];
+
+    let hasErrors = false;
```
don't need this, `diags.length > 0`

BarryNolte replied:
The return uses the hasErrors var in order to avoid multiple return paths.

alixander commented on server/src/d2Ast.ts:
```diff
@@ -0,0 +1,202 @@
+  }
+
+  /**
+   * Returns the list of links and imports to they can
```
typo

alixander commented on server/src/d2Ast.ts:
```diff
@@ -0,0 +1,202 @@
+
+      for (const node of this.nodes) {
+        // Edges
+        //
```
unnecessary line

alixander commented on server/src/d2Ast.ts:
is it mistake that there's `this.references` and `this.References`?

alixander commented on server/src/d2Ast.ts:
```diff
@@ -0,0 +1,202 @@
+   */
+  private doNodes(nodes: LSPAny[]) {
+    for (const node of nodes ?? []) {
+      // console.log("---------------------------------");
```
remove debug lines

alixander commented on server/src/d2Ast.ts:
```diff
@@ -0,0 +1,202 @@
+}
+
+/**
+ ***********************
```
what's the point of these

BarryNolte replied:
I had some test data after this and was trying to make it obvious, but its all gone now.

alixander commented on server/src/d2Ast.ts:
```diff
@@ -0,0 +1,202 @@
+   * Go through all the ast's node objects and break them
+   * apart to be used by the language server
+   */
+  private doNodes(nodes: LSPAny[]) {
```
better name please

alixander reviewed (changes requested):
looks like there's constant stream of errors mid-typing.

https://github.com/terrastruct/d2-vscode/assets/3120367/1a546e76-81dc-4c30-82b5-7ae4e0c4d8d4

-------

I reviewed 1 file. I like to test functionality while I review -- could you please update `README.md` to list the set of functionality that the LSP brings?

BarryNolte commented:
I checked out d2:master and it appears to be behind my d2:LSP-Fixes branch.  If you can checkout upstream/master, it should have my changes and the error spew will go away.  All new features are described in the PR Entry #117

BarryNolte reviewed (commented):
All issues resolved

alixander reviewed (changes requested):
Was able to test it all finally, feels really nice, especially the autocomplete list and import/link goto's 🚀 . 

---------

The implementation is a little flaky though. You want to defer as much of the interpretation of D2 to D2. You're already deferring parsing, but I see you've implemented finding references. This is actually quite cumbersome with many edge cases, for example, D2's underscore `_` mechanic that lets you reference parents and such. D2 actually has an an IR (intermediate representation). Though it's not specifically made for LSPs, it's designed with LSPs in mind from the start.

https://github.com/terrastruct/d2/tree/master/d2ir

Here's an example output of what d2ir spits out for `x.y`: https://github.com/terrastruct/d2/blob/master/testdata/d2ir/TestCompile/fields/primary/nested.exp.json

As you can see, it's got the list of fields and edges, and for each of those, all of its references.

----------

1. It seems to treat URLs as files.

https://github.com/terrastruct/d2-vscode/assets/3120367/c40f0a91-c02f-4809-ab1c-82fe1f652320

2,3,4. It works shallowly but it doesn't distinguish that a nested obj is different than one a level above.

<img width="355" alt="Screenshot 2024-02-21 at 1 58 49 PM" src="https://github.com/terrastruct/d2-vscode/assets/3120367/874a325e-390c-4d4e-b7f3-a7db6480ba97">

<img width="1295" alt="Screenshot 2024-02-21 at 1 58 58 PM" src="https://github.com/terrastruct/d2-vscode/assets/3120367/23a6983d-32d3-4279-8696-bf17871fbbe6">

5b. Is there any way for the list for "shape" to persist after a space is pressed after the colon?

https://github.com/terrastruct/d2-vscode/assets/3120367/5b4aff08-cda8-4c1a-ae52-79763221d16e

BarryNolte commented:
When I started this, I was passing the IR to the server, but then I figured out how to synthesize what I needed from the AST.  However, you found a limit that I should go back to using the IR for references and nested objects.  I'll work on that.

The autocomplete of ':' vs. ': ', I also thought was a good idea, and had it implemented.  It didn't quite feel right, at least to me.  So I tried a bunch of different languages, and every one of them auto completes on the token delimiter ('.', ':', etc).  The vscode autocomplete system is also geared toward this implementation. But, since I know I want a space in there, when you pick an item to autocompletion, it will put the space in.

alixander commented:
Can we split out the smaller ones from this large PR and have this just be focused on the language server? @BarryNolte
